src/strategies/base.ts

```diff
@@ -450,16 +448,27 @@ export abstract class BaseStrategy implements Strategy {
     return Version.parse('1.0.0');
   }
 
+  /**
+   * Adds a given file path to the strategy path.
+   * @param {string} file Desired file path.
+   * @returns {string} The file relative to the strategy.
+   * @throws {Error} If the file path contains relative pathing characters, i.e. ../
+   */
   protected addPath(file: string) {
-    if (this.path === ROOT_PROJECT_PATH) {
-      return file;
+    // There is no strategy path to join, the strategy is at the root, or the
+    // file is at the root (denoted by a leading slash or tilde)
+    if (!this.path || this.path === ROOT_PROJECT_PATH || /^~?\//.test(file)) {
+      file = file.replace(/^(~|\/)*\//, '');
+    }
+    // Otherwise, the file is relative to the strategy path
+    else {
+      file = `${this.path}/${file}`;
     }
-    file = file.replace(/^[/\\]/, '');
-    if (this.path === undefined) {
-      return file;
-    } else {
-      const path = this.path.replace(/[/\\]$/, '');
-      return `${path}/${file}`;
+    // Ensure the file path does not escape the workspace
+    if (/(\.{2}|^~|^\/*)+\//.test(file)) {
```

Comment:
Any of the below sequences in a file path are now rejected -

![image](https://user-images.githubusercontent.com/7577851/150670356-73d1c26e-4374-4ca3-929e-0f9982cc5b8e.png)